File: libaurum/src/PartialMatch.cc
/*
 * Copyright (c) 2021 Samsung Electronics Co., Ltd All Rights Reserved
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *               http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

#include "Aurum.h"

#include <iostream>
#include <set>
#include <regex>
#include <sstream>

using namespace Aurum;

bool PartialMatch::checkCriteria(const std::string textA, const std::string textB, const bool textPartialMatch)
{
    if (textB.empty()) return true;

    bool rst;
    if (textPartialMatch) {
        if (textB.find(textA) != std::string::npos) rst = false;
        else rst = true;
    }
    else {
        if (!textA.compare(textB)) rst = false;
        else rst = true;
    }

    return rst;
}

bool PartialMatch::checkCriteria(const bool boolA, const bool boolB)
{
    return boolA != boolB;
}

std::string PartialMatch::debugPrint()
{
    return mSelector->description();
}

bool PartialMatch::checkCriteria(const std::shared_ptr<UiSelector> selector,
                                 const std::shared_ptr<AccessibleNode> node)
{
    if (selector->mMatchText || selector->mMatchTextPartialMatch) {
        node->updateName();
        if (selector->mMatchText && checkCriteria(selector->mText, node->getText(), 0)) return false;
        if (selector->mMatchTextPartialMatch && checkCriteria(selector->mTextPartialMatch, node->getText(), 1)) return false;
    }
    if (selector->mMatchId) {
        node->updateUniqueId();
        if (checkCriteria(selector->mId, node->getId(), 0)) return false;
    }
    if (selector->mMatchType || selector->mMatchAutomationId || selector->mMatchStyle) {
        node->updateAttributes();
        if (selector->mMatchAutomationId && checkCriteria(selector->mAutomationId, node->getAutomationId(), 0)) return false;
        if (selector->mMatchType && checkCriteria(selector->mType, node->getType(), 0)) return false;
        if (selector->mMatchStyle && checkCriteria(selector->mStyle, node->getStyle(), 0)) return false;
    }
    if (selector->mMatchPkg) {
        node->updateApplication();
         if (checkCriteria(selector->mPkg, node->getPkg(), 0)) return false;
    }
    if (selector->mMatchRole) {
        node->updateRoleName();
        if (checkCriteria(selector->mRole, node->getRole(), 0)) return false;
    }
    if (selector->mMatchChecked && checkCriteria(selector->mIschecked, node->isChecked())) return false;
    if (selector->mMatchCheckable && checkCriteria(selector->mIscheckable, node->isCheckable())) return false;
    if (selector->mMatchClickable && checkCriteria(selector->mIsclickable, node->isClickable())) return false;
    if (selector->mMatchEnabled && checkCriteria(selector->mIsenabled, node->isEnabled())) return false;
    if (selector->mMatchFocused && checkCriteria(selector->mIsfocused, node->isFocused())) return false;
    if (selector->mMatchFocusable && checkCriteria(selector->mIsfocusable, node->isFocusable())) return false;
    if (selector->mMatchScrollable && checkCriteria(selector->mIsscrollable, node->isScrollable())) return false;
    if (selector->mMatchSelected && checkCriteria(selector->mIsselected, node->isSelected())) return false;
    if (selector->mMatchShowing && checkCriteria(selector->mIsshowing, node->isShowing())) return false;
    if (selector->mMatchActive && checkCriteria(selector->mIsactive, node->isActive())) return false;
    if (selector->mMatchVisible && checkCriteria(selector->mIsvisible, node->isVisible())) return false;
    if (selector->mMatchSelectable && checkCriteria(selector->mIsselectable, node->isSelectable())) return false;

    return true;
}

PartialMatch::PartialMatch() : mSelector{nullptr}, mDepth{-1}, mPartialMatches{}
{
}

PartialMatch::PartialMatch(const std::shared_ptr<UiSelector> selector, const int absDepth)
    : mSelector{selector}, mDepth{absDepth}, mPartialMatches{}
{
}

std::shared_ptr<PartialMatch> PartialMatch::accept(const std::shared_ptr<AccessibleNode> node,
                                                   const std::shared_ptr<UiSelector> selector,
                                                   int index, int depth)
{
    return PartialMatch::accept(node, selector, index, depth, depth);
}

std::shared_ptr<PartialMatch> PartialMatch::accept(const std::shared_ptr<AccessibleNode> node,
                                                   const std::shared_ptr<UiSelector> selector,
                                                   int index, int absoluteDepth,
                                                   int relativeDepth)
{
    PartialMatch *match = nullptr;

    if ((selector->mMinDepth && relativeDepth < selector->mMinDepth) ||
        (selector->mMaxDepth && relativeDepth > selector->mMaxDepth)) {
        return std::shared_ptr<PartialMatch>(nullptr);
    }
    if (PartialMatch::checkCriteria(selector, node))
        match = new PartialMatch(selector, absoluteDepth);
    return std::shared_ptr<PartialMatch>(match);
}

void PartialMatch::update(
    const std::shared_ptr<AccessibleNode> node, int index, int depth,
    std::list<std::shared_ptr<PartialMatch>> &partialMatches)
{
    for (auto &childSelector : mSelector->mChild) {
        auto match = PartialMatch::accept(node, childSelector, index, depth,
                                          depth - mDepth);
        if (match) {
            mPartialMatches.push_back(match);
            partialMatches.push_front(match);
        }
    }
}

bool PartialMatch::finalizeMatch()
{
    std::set<std::shared_ptr<UiSelector>> matches;
    for (auto &match : mPartialMatches) {
        if (match->finalizeMatch()) {
            matches.insert(match->mSelector);
        }
    }

    for (auto &sel : mSelector->mChild) {
        if (!matches.count(sel)) return false;
    }
    return true;
}

File: AurumSharp/AurumSharp/PartialMatch.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace SharpTest.libsharptest
{
    internal class PartialMatch
    {
        private UiSelector mSelector = null;

        private int mDepth = -1;

        private LinkedList<PartialMatch> mPartialMatches = new LinkedList<PartialMatch>();

        public PartialMatch(UiSelector selector, int depth)
        {
            this.mSelector = selector;
            this.mDepth = depth;
        }

        internal void update(AccessibleNode root, int index, int depth, LinkedList<PartialMatch> partialMatches)
        {
            foreach (var childSelector in mSelector.mChild)
            {
                var match = PartialMatch.accept(root, childSelector, index, depth,
                                                  depth - mDepth);
                if (match != null)
                {
                    mPartialMatches.AddLast(match);
                    partialMatches.AddFirst(match);
                }
            }
        }

        internal static PartialMatch accept(AccessibleNode root, UiSelector mSelector, int index, int depth)
        {
            return PartialMatch.accept(root, mSelector, index, depth, depth);
        }

        static PartialMatch accept(AccessibleNode node, UiSelector selector,
                                                   int index, int absoluteDepth,
                                                   int relativeDepth)
        {
            PartialMatch match = null;

            if ((selector.mMinDepth.HasValue && (relativeDepth < (selector.mMinDepth.Value))) ||
                (selector.mMaxDepth.HasValue && (relativeDepth > (selector.mMaxDepth.Value))))
            {
                return null;
            }
            if (PartialMatch.checkCriteria(selector, node)) match = new PartialMatch(selector, absoluteDepth);
            return match;
        }

        private static bool checkCriteria(UiSelector selector, AccessibleNode node)
        {
            if (selector == null || node == null) return false;

            if (checkCriteria(selector.mText, node.getText(), selector.mMatchText)) return true;
            if (checkCriteria(selector.mId, node.getId(), selector.mMatchId)) return true;
            if (checkCriteria(selector.mAutomationId, node.getAutomationId(), selector.mMatchAutomationId)) return true;
            if (checkCriteria(selector.mType, node.getType(), selector.mMatchType)) return true;
            if (checkCriteria(selector.mStyle, node.getStyle(), selector.mMatchStyle)) return true;
            if (checkCriteria(selector.mPkg, node.getPkg(), selector.mMatchPkg)) return true;
            if (checkCriteria(selector.mRole, node.getRole(), selector.mMatchRole)) return true;
            if (checkCriteria(selector.mIschecked, node.isChecked())) return true;
            if (checkCriteria(selector.mIscheckable, node.isCheckable())) return true;
            if (checkCriteria(selector.mIsclickable, node.isClickable())) return true;
            if (checkCriteria(selector.mIsenabled, node.isEnabled())) return true;
            if (checkCriteria(selector.mIsfocused, node.isFocused())) return true;
            if (checkCriteria(selector.mIsfocusable, node.isFocusable())) return true;
            if (checkCriteria(selector.mIsscrollable, node.isScrollable())) return true;
            if (checkCriteria(selector.mIsselected, node.isSelected())) return true;
            if (checkCriteria(selector.mIsshowing, node.isShowing())) return true;
            if (checkCriteria(selector.mIsactive, node.isActive())) return true;
            if (checkCriteria(selector.mIsvisible, node.isVisible())) return true;
            if (checkCriteria(selector.mIsselectable, node.isSelectable())) return true;

            return false;
        }

        private static bool checkCriteria(string textA, string textB, bool? mMatchText)
        {
            if (textA == null || textA.Length == 0 || mMatchText.HasValue == false) return false;
            
            Regex re = new Regex(textA);
            bool rst = (re.IsMatch(textB) == mMatchText);
            return rst;
        }

        private static bool checkCriteria(string textA, string textB)
        {
            return checkCriteria(textA, textB, true);
        }

        private static bool checkCriteria(bool? boolA, bool boolB)
        {
            if (!boolA.HasValue) return false;
            return boolA == boolB; 
        }

        internal bool finalizeMatch()
        {
            HashSet<UiSelector> matches = new HashSet<UiSelector>();
            
            foreach (var match in mPartialMatches)
            {
                if (match.finalizeMatch())
                {
                    matches.Add(match.mSelector);
                }
            }

            foreach (var sel in mSelector.mChild)
            {
                if (!matches.Contains(sel)) return false;
            }
            return true;
        }
    }
}